In [1]:
import numpy as np
import seaborn as sbs
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import f_oneway
import warnings
warnings.filterwarnings('ignore')

In [2]:
accident = pd.read_csv('datasets\\accident_data.csv')

# accident

In [3]:
accident

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [4]:
accident.describe()

Latitude      Longitude  Number_of_Casualties  Number_of_Vehicles
count  660654.000000  660653.000000         660679.000000       660679.000000
mean       52.553866      -1.431210              1.357040            1.831255
std         1.406922       1.383330              0.824847            0.715269
min        49.914430      -7.516225              1.000000            1.000000
25%        51.490690      -2.332291              1.000000            1.000000
50%        52.315641      -1.411667              1.000000            2.000000
75%        53.453452      -0.232869              1.000000            2.000000
max        60.757544       1.762010             68.000000           32.000000

In [5]:
accident.isnull().sum()

Index                          0
Accident_Severity              0
Accident Date                  0
Latitude                      25
Light_Conditions               0
District Area                  0
Longitude                     26
Number_of_Casualties           0
Number_of_Vehicles             0
Road_Surface_Conditions      726
Road_Type                   4520
Urban_or_Rural_Area           15
Weather_Conditions         14128
Vehicle_Type                   0
dtype: int64

In [6]:
accident['Latitude'] = accident['Latitude'].fillna(accident['Latitude'].mode()[0])
accident['Longitude'] = accident['Longitude'].fillna(accident['Longitude'].mode()[0])
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].fillna(accident['Road_Surface_Conditions'].mode()[0])
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].fillna(accident['Urban_or_Rural_Area'].mode()[0])
accident['Weather_Conditions'] = accident['Weather_Conditions'].fillna('unaccounted')
accident['Road_Type'] = accident['Road_Type'].fillna('unknown')

In [7]:
accident.isnull().sum()

Index                      0
Accident_Severity          0
Accident Date              0
Latitude                   0
Light_Conditions           0
District Area              0
Longitude                  0
Number_of_Casualties       0
Number_of_Vehicles         0
Road_Surface_Conditions    0
Road_Type                  0
Urban_or_Rural_Area        0
Weather_Conditions         0
Vehicle_Type               0
dtype: int64

In [8]:
accident.dtypes

Index                       object
Accident_Severity           object
Accident Date               object
Latitude                   float64
Light_Conditions            object
District Area               object
Longitude                  float64
Number_of_Casualties         int64
Number_of_Vehicles           int64
Road_Surface_Conditions     object
Road_Type                   object
Urban_or_Rural_Area         object
Weather_Conditions          object
Vehicle_Type                object
dtype: object

In [9]:
accident['Accident_Severity'] = accident['Accident_Severity'].astype('category')
accident['Light_Conditions'] = accident['Light_Conditions'].astype('category')
accident['District Area'] = accident['District Area'].astype('category')
accident['Road_Surface_Conditions'] = accident['Road_Surface_Conditions'].astype('category')
accident['Road_Type'] = accident['Road_Type'].astype('category')
accident['Urban_or_Rural_Area'] = accident['Urban_or_Rural_Area'].astype('category')
accident['Vehicle_Type'] = accident['Vehicle_Type'].astype('category')

In [10]:
accident['Accident Date'] = pd.to_datetime(accident['Accident Date'], dayfirst = True, errors = 'coerce')

In [11]:
accident.dtypes

Index                              object
Accident_Severity                category
Accident Date              datetime64[ns]
Latitude                          float64
Light_Conditions                 category
District Area                    category
Longitude                         float64
Number_of_Casualties                int64
Number_of_Vehicles                  int64
Road_Surface_Conditions          category
Road_Type                        category
Urban_or_Rural_Area              category
Weather_Conditions                 object
Vehicle_Type                     category
dtype: object

<h1>1. Which area is more likely to have an accident, urban or rural?</h1>

In [12]:
accident['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          421678
Rural          238990
Unallocated        11
Name: count, dtype: int64

In [13]:
count = accident['Urban_or_Rural_Area'].value_counts()
total = count.sum()
percent = (count / total) * 100
percent

Urban_or_Rural_Area
Urban          63.824944
Rural          36.173391
Unallocated     0.001665
Name: count, dtype: float64

In [14]:
accident['Urban_or_Rural_Area'].value_counts().idxmax()

'Urban'

<h2>Insight 1: the data show that Urban is more likely to have an accident with 64% of the total accident being record.</h2>


<h1></h1>

<h1>2.Our data shows that Urban is more likely to have an accident, but what about casualties, which Area has more casualty?</h1>

In [15]:
accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()

Urban_or_Rural_Area
Rural          1.479204
Unallocated    1.181818
Urban          1.287807
Name: Number_of_Casualties, dtype: float64

In [16]:
avg = accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean()
total = avg.sum()
percent = (avg / total) * 100
percent

Urban_or_Rural_Area
Rural          37.459306
Unallocated    29.928315
Urban          32.612379
Name: Number_of_Casualties, dtype: float64

In [17]:
accident.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].mean().idxmax()

'Rural'

<h3>Insight 2: Now as we can See even though Urban is more likely to have an accident, the data shows that Rural have more casualties with 37% of the total casualties being record.</h3>

<h1></h1>

<h1>3. now let see about the other concept, let look at the Severity, which Severity is more likely to have casualties</h1>

In [18]:
accident.groupby(['Accident_Severity','Urban_or_Rural_Area'])['Number_of_Casualties'].mean().idxmax()

('Fatal', 'Rural')

In [19]:
avg = accident.groupby(['Accident_Severity','Urban_or_Rural_Area'])['Number_of_Casualties'].mean()
total = avg.sum()
percent = (avg / total) * 100
percent

Accident_Severity  Urban_or_Rural_Area
Fatal              Rural                  18.145282
                   Unallocated                  NaN
                   Urban                  13.445978
Serious            Rural                  14.572156
                   Unallocated             8.662041
                   Urban                  11.344510
Slight             Rural                  12.325838
                   Unallocated            10.394449
                   Urban                  11.109746
Name: Number_of_Casualties, dtype: float64

<h3>Insight 3:now this data shows that, Fatal Accident in Rural Area have more Casualties involve through out the whole record, with 18% of the total record was a fatal accident in Rural Area. Which is obvious since Fatal accident tends to have more casualties than the other other accident severities, what am I even thinking of making this? </h3>

<h1></h1>

<h1>4. Now we saw all of that Fatal Accident in Rural Area have more Casualties, lets look at the other factor</h1>
<h2>How about which Vehicle? Which Vehicle in Urban And Rural Area are most likely to have accident and Casualties involve?</h2>

In [20]:
urban = accident[accident['Urban_or_Rural_Area']=='Urban']
rural = accident[accident['Urban_or_Rural_Area']=='Rural']

In [21]:
rural.groupby(['Vehicle_Type','Accident_Severity'])['Number_of_Casualties'].mean().idxmax()

('Minibus (8 - 16 passenger seats)', 'Fatal')

In [22]:
urban.groupby(['Vehicle_Type','Accident_Severity'])['Number_of_Casualties'].mean().idxmax()

('Goods over 3.5t. and under 7.5t', 'Fatal')

In [23]:
avg = rural.groupby(['Vehicle_Type','Accident_Severity'])['Number_of_Casualties'].mean()
total = avg.sum()
percent = (avg / total) * 100
percent.nlargest(1)

Vehicle_Type                      Accident_Severity
Minibus (8 - 16 passenger seats)  Fatal                4.514994
Name: Number_of_Casualties, dtype: float64

In [24]:
avg = urban.groupby(['Vehicle_Type','Accident_Severity'])['Number_of_Casualties'].mean()
total = avg.sum()
percent = (avg / total) * 100
percent.nlargest(1)

Vehicle_Type                     Accident_Severity
Goods over 3.5t. and under 7.5t  Fatal                3.379246
Name: Number_of_Casualties, dtype: float64

<h3>Insight 4: Now as we can See the Minibus has the highest rate of casualties being record as 4.5% of the total casualties in Rural Area is involved an Fatal Accident.</h3>
<h3>And in Urban Area we can also See that Vehicle that is Goods over 3.5t. and under 7.5t has the highest rate of casualties being record as 3.4% of the total casualties in Urban area is involved in a fatal accident.</h3>

<h4>There for conclude that Minibus (8 - 16 passenger seats) and Vehicle that is Goods over 3.5t. and under 7.5t is more likely to have an accident.</h4>

<h1></h1>

<h1>5.Now How about the total on all of the accident record, which Vehicle has more Casualties involve in urban and rural?</h1>

In [25]:
urban.groupby('Vehicle_Type')['Number_of_Casualties'].mean().nlargest(3)

Vehicle_Type
Data missing or out of range    1.500000
Motorcycle over 500cc           1.296467
Taxi/Private hire car           1.296391
Name: Number_of_Casualties, dtype: float64

In [26]:
rural.groupby('Vehicle_Type')['Number_of_Casualties'].mean().nlargest(3)

Vehicle_Type
Ridden horse             2.000000
Pedal cycle              1.514286
Taxi/Private hire car    1.501817
Name: Number_of_Casualties, dtype: float64

In [27]:
avg = urban.groupby('Vehicle_Type')['Number_of_Casualties'].mean()
total = avg.sum()
percent = (avg / total) * 100
percent

Vehicle_Type
Agricultural vehicle                     6.286426
Bus or coach (17 or more pass seats)     6.242696
Car                                      6.282796
Data missing or out of range             7.313720
Goods 7.5 tonnes mgw and over            6.306507
Goods over 3.5t. and under 7.5t          6.257546
Minibus (8 - 16 passenger seats)         6.309876
Motorcycle 125cc and under               6.171621
Motorcycle 50cc and under                6.201142
Motorcycle over 125cc and up to 500cc    6.262612
Motorcycle over 500cc                    6.321332
Other vehicle                            6.287832
Pedal cycle                              6.296326
Ridden horse                             4.875814
Taxi/Private hire car                    6.320960
Van / Goods 3.5 tonnes mgw or under      6.262793
Name: Number_of_Casualties, dtype: float64

In [28]:
avg = urban.groupby('Vehicle_Type')['Number_of_Casualties'].mean()
total = avg.sum()
percent = (avg / total) * 100
percent

Vehicle_Type
Agricultural vehicle                     6.286426
Bus or coach (17 or more pass seats)     6.242696
Car                                      6.282796
Data missing or out of range             7.313720
Goods 7.5 tonnes mgw and over            6.306507
Goods over 3.5t. and under 7.5t          6.257546
Minibus (8 - 16 passenger seats)         6.309876
Motorcycle 125cc and under               6.171621
Motorcycle 50cc and under                6.201142
Motorcycle over 125cc and up to 500cc    6.262612
Motorcycle over 500cc                    6.321332
Other vehicle                            6.287832
Pedal cycle                              6.296326
Ridden horse                             4.875814
Taxi/Private hire car                    6.320960
Van / Goods 3.5 tonnes mgw or under      6.262793
Name: Number_of_Casualties, dtype: float64

In [29]:
urban_avg = urban.groupby('Vehicle_Type')['Number_of_Casualties'].mean()
rural_avg = rural.groupby('Vehicle_Type')['Number_of_Casualties'].mean()

urban_total = urban_avg.sum()
rural_total = rural_avg.sum()

urban_percent = (urban_avg / urban_total) * 100
rural_percent = (rural_avg / rural_total) * 100
urban_table = pd.DataFrame({'Urban Percentage': urban_percent}).sort_values(by='Urban Percentage', ascending=False)
rural_table = pd.DataFrame({'Rural Percentage': rural_percent}).sort_values(by='Rural Percentage', ascending=False)
display(urban_table)
display(rural_table)


Urban Percentage
Vehicle_Type                                           
Data missing or out of range                   7.313720
Motorcycle over 500cc                          6.321332
Taxi/Private hire car                          6.320960
Minibus (8 - 16 passenger seats)               6.309876
Goods 7.5 tonnes mgw and over                  6.306507
Pedal cycle                                    6.296326
Other vehicle                                  6.287832
Agricultural vehicle                           6.286426
Car                                            6.282796
Van / Goods 3.5 tonnes mgw or under            6.262793
Motorcycle over 125cc and up to 500cc          6.262612
Goods over 3.5t. and under 7.5t                6.257546
Bus or coach (17 or more pass seats)           6.242696
Motorcycle 50cc and under                      6.201142
Motorcycle 125cc and under                     6.171621
Ridden horse                                   4.875814

Rural Percentage
Vehicle_Type                                           
Ridden horse                                   8.857738
Pedal cycle                                    6.706573
Taxi/Private hire car                          6.651353
Goods over 3.5t. and under 7.5t                6.645288
Van / Goods 3.5 tonnes mgw or under            6.560964
Car                                            6.558900
Bus or coach (17 or more pass seats)           6.545402
Motorcycle over 500cc                          6.540704
Motorcycle 125cc and under                     6.506197
Goods 7.5 tonnes mgw and over                  6.456249
Motorcycle 50cc and under                      6.445557
Motorcycle over 125cc and up to 500cc          6.437926
Agricultural vehicle                           6.384133
Minibus (8 - 16 passenger seats)               6.359560
Other vehicle                                  6.343456
Data missing or out of range                        NaN

<h3>Insight 5: Based on our data, it is shown that in Urban Area, Motorcycle over 500cc and Taxi/Private hire car have more Casualties Total being record</h3>
<h3>And in Rural Area Pedal cycle and Taxi/Private hire car have more Casualties Total being record </h3>

<h1></h1>

<h1>6. Since we are Talking about Vehicle, how about looking at which Vehicle cause more, fatal accident, Serious accident and Slight accident</h1>

In [30]:
Fatal = accident[accident['Accident_Severity']=='Fatal']
Serious = accident[accident['Accident_Severity']=='Serious']
Slight = accident[accident['Accident_Severity']=='Slight']

In [31]:
Fatal['Vehicle_Type'].value_counts().nlargest(5)

Vehicle_Type
Car                                     6577
Van / Goods 3.5 tonnes mgw or under      467
Motorcycle over 500cc                    339
Bus or coach (17 or more pass seats)     325
Goods 7.5 tonnes mgw and over            216
Name: count, dtype: int64

In [32]:
Serious['Vehicle_Type'].value_counts().nlargest(5)

Vehicle_Type
Car                                     66461
Van / Goods 3.5 tonnes mgw or under      4554
Motorcycle over 500cc                    3457
Bus or coach (17 or more pass seats)     3373
Goods 7.5 tonnes mgw and over            2321
Name: count, dtype: int64

In [33]:
Slight['Vehicle_Type'].value_counts().nlargest(5)

Vehicle_Type
Car                                     424954
Van / Goods 3.5 tonnes mgw or under      29139
Bus or coach (17 or more pass seats)     22180
Motorcycle over 500cc                    21861
Goods 7.5 tonnes mgw and over            14770
Name: count, dtype: int64

<h3>Insight 6: Car have serious Issue, it is shown that In all Accident Severity, Car is more likely to have an accident, as well as Van / Goods 3.5 tonnes mgw or under </h3>

In [34]:
accident['Accident_Severity'].value_counts()

Accident_Severity
Slight     563801
Serious     88217
Fatal        8661
Name: count, dtype: int64

In [35]:
accident['Accident_Severity'].value_counts().idxmax()

'Slight'

<h2>Insight 6: The most common accident severity to cause is Slight Accident</h2>

<h1></h1>

<h1>7. Let stop on Vehicle for now, lets talk about Road Type now, Which Road Type as well as their conditions cause more fatal, serious, and slight accident? </h1>

In [36]:
Fatal.groupby(['Road_Type', 'Road_Surface_Conditions']).size().nlargest(5)

Road_Type           Road_Surface_Conditions
Single carriageway  Dry                        4338
                    Wet or damp                1993
Dual carriageway    Dry                        1217
                    Wet or damp                 551
Single carriageway  Frost or ice                153
dtype: int64

In [37]:
Serious.groupby(['Road_Type', 'Road_Surface_Conditions']).size().nlargest(5)

Road_Type           Road_Surface_Conditions
Single carriageway  Dry                        48926
                    Wet or damp                18952
Dual carriageway    Dry                         8104
                    Wet or damp                 3233
Roundabout          Dry                         2640
dtype: int64

In [38]:
Slight.groupby(['Road_Type', 'Road_Surface_Conditions']).size().nlargest(5)

Road_Type           Road_Surface_Conditions
Single carriageway  Dry                        279961
                    Wet or damp                117798
Dual carriageway    Dry                         56951
Roundabout          Dry                         28034
Dual carriageway    Wet or damp                 25749
dtype: int64

<h3>Insight 7: Hhhmmmm this is interesting, I think the City or the Country need to do something about those results, since it shown that Single and Dual carriageway cause too much accident when they are Dry or 'Wet or damp'. either Fatal or Serious or Slight Accident, they mostly cause the accidents</h3>

<h1></h1>

<h1>8. And in urban and Rural which Road Types as well as their Conditions Cause more Accident?</h1>

In [39]:
urban.groupby(['Road_Type', 'Road_Surface_Conditions']).size().nlargest(10)

Road_Type           Road_Surface_Conditions
Single carriageway  Dry                        234694
                    Wet or damp                 81073
Dual carriageway    Dry                         35062
Roundabout          Dry                         19808
Dual carriageway    Wet or damp                 14394
One way street      Dry                          9270
Roundabout          Wet or damp                  8022
Single carriageway  Frost or ice                 5956
One way street      Wet or damp                  2869
unknown             Dry                          2646
dtype: int64

In [40]:
rural.groupby(['Road_Type', 'Road_Surface_Conditions']).size().nlargest(10)

Road_Type           Road_Surface_Conditions
Single carriageway  Dry                        98526
                    Wet or damp                57666
Dual carriageway    Dry                        31210
                    Wet or damp                15139
Roundabout          Dry                        10976
Single carriageway  Frost or ice                8962
Roundabout          Wet or damp                 4186
Slip road           Dry                         2803
Single carriageway  Snow                        2325
Dual carriageway    Frost or ice                1498
dtype: int64

<h3>Insight 8: Well that's to be expected, Single and Dual carriageway when they are either Dry or 'Wet or damp' Causes more accident through out the whole record.</h3>

<h1></h1>

<h1>9. Since we have those record where we saw that Single and Dual carriageway cause more accident than the rest of the other road, let see which type of Vehicle is more likely to have an accident on those road </h1>

In [41]:
single = accident[accident['Road_Type']=='Single carriageway']
dual = accident[accident['Road_Type']=='Dual carriageway']

In [42]:
single['Vehicle_Type'].value_counts().nlargest(5)

Vehicle_Type
Car                                     371028
Van / Goods 3.5 tonnes mgw or under      25460
Bus or coach (17 or more pass seats)     19090
Motorcycle over 500cc                    19050
Goods 7.5 tonnes mgw and over            12915
Name: count, dtype: int64

In [43]:
dual['Vehicle_Type'].value_counts().nlargest(5)

Vehicle_Type
Car                                     74820
Van / Goods 3.5 tonnes mgw or under      5107
Bus or coach (17 or more pass seats)     4043
Motorcycle over 500cc                    3930
Goods 7.5 tonnes mgw and over            2631
Name: count, dtype: int64

<h3>insight 9: Car really have some issue, it is shown that on single and dual carriageway car followed by van and bus is more likely to cause an accident.</h3>

<h1></h1>

<h1>10. Let's stop on Car and start looking at District Area</h1>
<h2>Which Disctrict is more likely to have an accident in urban and rural</h2>

In [44]:
urban = accident[accident['Urban_or_Rural_Area'] == 'Urban']
urban_count = urban['District Area'].value_counts()
urban_total = urban_count.sum()
urban_percent = (urban_count / urban_total) * 100

urban_df = pd.DataFrame({'District Area': urban_count.index, 'Accident Count': urban_count.values, 'Percentage': urban_percent.values})
urban_df = urban_df.head(10).reset_index(drop=True)

rural = accident[accident['Urban_or_Rural_Area'] == 'Rural']
rural_count = rural['District Area'].value_counts()
rural_total = rural_count.sum()
rural_percent = (rural_count / rural_total) * 100

rural_df = pd.DataFrame({'District Area': rural_count.index, 'Accident Count': rural_count.values, 'Percentage': rural_percent.values})
rural_df = rural_df.head(10).reset_index(drop=True)

print("Top 10 Urban Districts with the Most Accidents:")
display(urban_df)

print("\nTop 10 Rural Districts with the Most Accidents:")
display(rural_df)

Top 10 Urban Districts with the Most Accidents:


District Area  Accident Count  Percentage
0        Birmingham           13357    3.167583
1             Leeds            7124    1.689441
2        Manchester            6577    1.559721
3       Westminster            5706    1.353165
4         Liverpool            5539    1.313562
5          Bradford            5416    1.284392
6         Sheffield            5248    1.244552
7      Glasgow City            4806    1.139732
8  Bristol, City of            4581    1.086374
9           Lambeth            4241    1.005744


Top 10 Rural Districts with the Most Accidents:


District Area  Accident Count  Percentage
0  East Riding of Yorkshire            2364    0.989163
1                  Cornwall            1992    0.833508
2                  Highland            1811    0.757772
3      South Cambridgeshire            1803    0.754425
4             Aberdeenshire            1781    0.745219
5                     Leeds            1774    0.742290
6                     Powys            1692    0.707979
7           Huntingdonshire            1673    0.700029
8              East Lindsey            1538    0.643542
9           Carmarthenshire            1527    0.638939

In [45]:
count.sum()

np.int64(660679)

<h3>Insight 10: As we can see the data shows that Birmingham was most likely to have an accident in urban area as 3% of the accident being record in urban area was in Birmingham</h3>
<h3> on the other hand in rural area, East Riding of Yorkshire was most likely to have an accident with 1% of the accident in the rural area happened on East Riding of Yorkshire</h3>

<h1></h1>

<h1>11. now, Which Disctrict area have the record with the most casualties involved? and Which is the District have the least number of casualties involved</h1>

In [46]:
total_casualties = accident['Number_of_Casualties'].sum()
print("Total Number of Casualties:", total_casualties)

Total Number of Casualties: 896568


In [47]:
num = accident.groupby('District Area')['Number_of_Casualties'].sum()
total = num.sum()
percent = (num / total) * 100
percent.nlargest(5)

District Area
Birmingham    2.082831
Leeds         1.387290
Manchester    1.052569
Bradford      1.033720
Liverpool     0.966909
Name: Number_of_Casualties, dtype: float64

In [48]:
num.nlargest(5)

District Area
Birmingham    18674
Leeds         12438
Manchester     9437
Bradford       9268
Liverpool      8669
Name: Number_of_Casualties, dtype: int64

In [49]:
accident.groupby('District Area')['Number_of_Casualties'].sum().nsmallest(10)

District Area
Clackmannanshire             115
Orkney Islands               154
Shetland Islands             202
Clackmannshire               204
London Airport (Heathrow)    217
Berwick-upon-Tweed           222
Teesdale                     222
Western Isles                249
South Shropshire             324
Alnwick                      351
Name: Number_of_Casualties, dtype: int64

In [50]:
num = accident.groupby('District Area')['Number_of_Casualties'].sum()

# Get top 5 and bottom 10 districts
top_5 = num.nlargest(5)
bottom_5 = num.nsmallest(5)

# Convert to DataFrame for better display
top_5_df = top_5.reset_index().rename(columns={'Number_of_Casualties': 'Total Casualties'})
bottom_5_df = bottom_5.reset_index().rename(columns={'Number_of_Casualties': 'Total Casualties'})

# Display tables
print("Top 5 Districts with the Most Casualties:")
display(top_5_df)

print("\nBottom 10 Districts with the Least Casualties:")
display(bottom_5_df)

Top 5 Districts with the Most Casualties:


District Area  Total Casualties
0    Birmingham             18674
1         Leeds             12438
2    Manchester              9437
3      Bradford              9268
4     Liverpool              8669


Bottom 10 Districts with the Least Casualties:


District Area  Total Casualties
0           Clackmannanshire               115
1             Orkney Islands               154
2           Shetland Islands               202
3             Clackmannshire               204
4  London Airport (Heathrow)               217

<h3>Insight 11: The data indicates that out of 89.7K total recorded casualties, 2% occurred in Birmingham, totaling 18,674 casualties in that district.</h3>
<h3>Meanwhile, Clackmannanshire recorded the fewest casualties, with only 115 throughout the entire dataset.</h3>

<h1></h1>

<h1>12. Which district have the most vehicle accident involve and what type of vehicle always cause an accident in that destrict (beside car)</h1>

In [51]:
veh_per_dist = accident.groupby('District Area')['Number_of_Vehicles'].sum()
top_dist = veh_per_dist.idxmax()
top_veh_count = veh_per_dist.max()

print(f"District with the most vehicles involved in accidents: {top_dist} ({top_veh_count} vehicles)")

dist_data = accident[accident['District Area'] == top_dist]

car = dist_data['Vehicle_Type'].value_counts().idxmax()
car_count = dist_data['Vehicle_Type'].value_counts().max()
car_pct = (car_count / dist_data.shape[0]) * 100

non_car = dist_data[dist_data['Vehicle_Type'] != 'Car']
non_car_type = non_car['Vehicle_Type'].value_counts().idxmax()
non_car_count = non_car['Vehicle_Type'].value_counts().max()
non_car_pct = (non_car_count / dist_data.shape[0]) * 100


print(f"The Car, known to be the most problematic vehicle, has a record in {top_dist}: {car} ({car_count} times, {car_pct:.2f}%)")
print(f"Most common vehicle type involved (besides cars) in {top_dist}: {non_car_type} ({non_car_count} times, {non_car_pct:.2f}%)")


District with the most vehicles involved in accidents: Birmingham (24869 vehicles)
The Car, known to be the most problematic vehicle, has a record in Birmingham: Car (9600 times, 71.16%)
Most common vehicle type involved (besides cars) in Birmingham: Van / Goods 3.5 tonnes mgw or under (785 times, 5.82%)


<h3>Insight 12: The data reveals that Birmingham has the highest number of car accidents, with cars accounting for 71% of all recorded incidents. Additionally, Vans/Goods vehicles (3.5 tonnes mgw or under) rank second but contribute to only 6% of the total accidents.</h3>

<h3></h3>

<h1></h1>

<h2>I think We had enough of cars record, let's talk about time and dates</h2>
<h1>13. How about looking at which year had the most accident happened through out the whole record, is there an improvement each year or the accident is just getting worse?</h1>

In [52]:
accident['Year'] = accident['Accident Date'].dt.year
accident['Year'].value_counts()

Year
2019.0    71867
2020.0    70163
2021.0    66172
2022.0    56805
Name: count, dtype: int64

<h3>Insight 13: The data shows a consistent decline in the number of accidents each year, indicating an overall improvement in road safety over time.</h3>

<h1></h1>

<h1>14. How about Per month, which month is have more accident happened or more likely to have an accident?</h1>

In [53]:
accident['Month'] = accident['Accident Date'].dt.month
accident['Month'].value_counts()

Month
11.0    24240
12.0    24156
10.0    23962
7.0     22939
9.0     22558
2.0     22264
6.0     22196
3.0     21824
5.0     21723
8.0     21106
4.0     19787
1.0     18252
Name: count, dtype: int64

<h3>Insight 14: The data shows that November and December the highest number of accidents, while January and April have the lowest, indicating seasonal variations in accident trends.</h3>

<h1></h1>

<h1>15. How about which district have more accident each year?</h1>

In [54]:
accident['Year'] = accident['Accident Date'].dt.year  
top_districts_per_year = {}
years = [2019, 2020, 2021, 2022]

for year in years:
    top_districts_per_year[year] = (
        accident[accident['Year'] == year]
        .groupby('District Area')
        .size()
        .nlargest(10)
    )

for year, data in top_districts_per_year.items():
    print(f"\nTop 10 Districts with Most Accidents in {year}:\n")
    print(data)
    print("-" * 50)



Top 10 Districts with Most Accidents in 2019:

District Area
Birmingham            1554
Leeds                  950
Glasgow City           721
Manchester             684
Liverpool              628
Sheffield              616
Bradford               614
Westminster            608
Bristol, City of       557
Edinburgh, City of     524
dtype: int64
--------------------------------------------------

Top 10 Districts with Most Accidents in 2020:

District Area
Birmingham            1438
Leeds                  921
Manchester             721
Glasgow City           676
Sheffield              600
Liverpool              597
Bradford               587
Westminster            579
Edinburgh, City of     547
Kirklees               524
dtype: int64
--------------------------------------------------

Top 10 Districts with Most Accidents in 2021:

District Area
Birmingham            1308
Leeds                  914
Manchester             690
Bradford               641
Sheffield              614
Glasgow Cit

<h3>Insight 15: From 2019 to 2022, Birmingham consistently recorded the highest number of accidents each year, making it the most accident-prone district. Leeds and Manchester also appeared among the top 3 districts with the most accidents in all four years, highlighting persistent accident hotspots in these urban areas.</h3>

<h1></h1>

<h1>16. How do seasonal changes affect the number of road accidents throughout the year? </h1>

In [55]:
def get_season(month):
    if month in [12, 1, 2]:
        return 'Winter'
    elif month in [3, 4, 5]:
        return 'Spring'
    elif month in [6, 7, 8]:
        return 'Summer'
    else:
        return 'Fall'

accident['Season'] = accident['Accident Date'].dt.month.apply(get_season)

In [56]:
accident['Season'].value_counts().reindex(['Winter', 'Spring', 'Summer', 'Fall'])

Season
Winter     64672
Spring     63334
Summer     66241
Fall      466432
Name: count, dtype: int64

<h3>Insight 16: The data reveals a significant seasonal pattern in accidents, with fall recording the highest number of incidents. This may be due to factors like reduced daylight, wet road conditions, and increased traffic activity. Meanwhile, winter, spring, and summer show relatively lower accident counts.</h3>

<h1></h1>

<h1>17. Do poor lighting conditions or hazardous road surfaces contribute to the high number of accidents in the fall season?</h1>

In [57]:
accident.groupby(['Season', 'Light_Conditions']).size().unstack()

Light_Conditions  Darkness - lighting unknown  Darkness - lights lit  \
Season                                                                 
Fall                                     4650                  93412   
Spring                                    431                   9184   
Summer                                    357                   5907   
Winter                                   1046                  20832   

Light_Conditions  Darkness - lights unlit  Darkness - no lighting  Daylight  
Season                                                                       
Fall                                 1815                   26740    339815  
Spring                                200                    2671     50848  
Summer                                163                    1780     58034  
Winter                                365                    6246     36183

In [58]:
accident.groupby(['Season', 'Road_Surface_Conditions']).size().unstack()

Road_Surface_Conditions     Dry  Flood over 3cm. deep  Frost or ice  Snow  \
Season                                                                      
Fall                     318106                   753          9466  3003   
Spring                    51722                    47           576   199   
Summer                    53218                   106             6    11   
Winter                    25501                   111          8469  2677   

Road_Surface_Conditions  Wet or damp  
Season                                
Fall                          135104  
Spring                         10790  
Summer                         12900  
Winter                         27914

In [59]:
accident.groupby(['Season', 'Weather_Conditions']).size().unstack()

Weather_Conditions  Fine + high winds  Fine no high winds  Fog or mist  Other  \
Season                                                                          
Fall                             5882              367079         2618  11113   
Spring                           1013               54346           93    870   
Summer                            235               55609           70    748   
Winter                           1424               43851          747   4419   

Weather_Conditions  Raining + high winds  Raining no high winds  \
Season                                                            
Fall                                6949                  59207   
Spring                               736                   4657   
Summer                               303                   7926   
Winter                              1627                   7906   

Weather_Conditions  Snowing + high winds  Snowing no high winds  unaccounted  
Season                                                                        
Fall                                 469                   3176         9939  
Spring                                47                    263         1309  
Summer                                 1                     21         1328  
Winter                               368                   2778         1552

<h3>Insight 17: Fall records the highest number of accidents, primarily due to wet road conditions and heavy rainfall rather than poor lighting. While darkness contributes, the data shows that most accidents occur on wet or damp roads, often during rainy weather. This suggests that slippery surfaces and reduced visibility from precipitation are key risk factors in the fall season.</h3>

<h1></h1>

<h1>18. Relationship Between Number of Vehicles and Accident Severity</h1>

In [60]:
accident.groupby(['Season', 'Accident_Severity'])['Number_of_Vehicles'].describe()

count      mean       std  min  25%  50%  75%  \
Season Accident_Severity                                                     
Fall   Fatal                6067.0  1.793803  1.044592  1.0  1.0  2.0  2.0   
       Serious             62418.0  1.681262  0.752348  1.0  1.0  2.0  2.0   
       Slight             397947.0  1.857064  0.700476  1.0  1.0  2.0  2.0   
Spring Fatal                 893.0  1.793953  0.964672  1.0  1.0  2.0  2.0   
       Serious              8660.0  1.668245  0.744683  1.0  1.0  2.0  2.0   
       Slight              53781.0  1.858407  0.700220  1.0  1.0  2.0  2.0   
Summer Fatal                 867.0  1.835063  1.023632  1.0  1.0  2.0  2.0   
       Serious              9177.0  1.705023  0.721184  1.0  1.0  2.0  2.0   
       Slight              56197.0  1.870367  0.677241  1.0  1.0  2.0  2.0   
Winter Fatal                 834.0  1.679856  0.909812  1.0  1.0  1.0  2.0   
       Serious              7962.0  1.635519  0.783760  1.0  1.0  2.0  2.0   
       Slight              55876.0  1.830285  0.723669  1.0  1.0  2.0  2.0   

                           max  
Season Accident_Severity        
Fall   Fatal              28.0  
       Serious            19.0  
       Slight             32.0  
Spring Fatal               8.0  
       Serious             8.0  
       Slight             14.0  
Summer Fatal              14.0  
       Serious             8.0  
       Slight             10.0  
Winter Fatal               7.0  
       Serious            12.0  
       Slight             14.0

<h1></h1>

<h1>19. Urban vs. Rural Accidents in Fall</h1>

In [61]:
accident[accident['Season'] == 'Fall']['Urban_or_Rural_Area'].value_counts()

Urban_or_Rural_Area
Urban          298502
Rural          167923
Unallocated         7
Name: count, dtype: int64

<h1></h1>

<h1>20. Accident Trends by Road Type</h1>

In [75]:
accident.groupby(['Road_Type', 'Season']).size().unstack()

Season                Fall  Spring  Summer  Winter
Road_Type                                         
Dual carriageway     70258    9253    9626   10287
One way street        9582    1311    1380    1286
Roundabout           31279    4169    4502    4042
Single carriageway  347209   47482   49526   47926
Slip road             4944     693     685     719
unknown               3160     426     522     412

<h1></h1>

In [69]:
accident.groupby('Road_Type')['Number_of_Casualties'].mean()

Road_Type
Dual carriageway      1.477279
One way street        1.192713
Roundabout            1.274891
Single carriageway    1.344666
Slip road             1.423661
unknown               1.248230
Name: Number_of_Casualties, dtype: float64

In [70]:
accident.groupby('Road_Type')['Number_of_Casualties'].mean().idxmax()

'Dual carriageway'

<h2>Insight 5: The Data Shows that Dual Carriageway Road is more likely to have casualty than the others, meaning that people either avoid the road or the City do something about the road like changing it?</h2>